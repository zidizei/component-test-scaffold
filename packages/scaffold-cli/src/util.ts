import baseGlob from "glob"
import path from "path"

export async function glob(patterns: Array<string>): Promise<string[]> {
    const pattern = (patterns.length > 0) ? `+(${patterns.join("|")}).*` : "*"

    return new Promise((resolve, reject) => {
        baseGlob(path.resolve(`./**/__scaffolds__/${pattern}.scaffold.js`), (err, matches) => {
            if (err) { return reject(err) }
            resolve(matches)
        })
    })
}
