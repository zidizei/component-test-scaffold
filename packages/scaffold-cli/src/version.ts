const {version: VERSION} = require("../package.json")

export default (): string => VERSION
