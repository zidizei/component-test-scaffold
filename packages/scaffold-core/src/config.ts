import urlUtil from "url"

import { ScaffoldOptions } from "."

const getFilenameFromModule = (parent: NodeModule | null) => parent && parent.filename

const getCasenameFromUrl = (url: string) => {
    const pathname = urlUtil.parse(url).pathname

    return (pathname === undefined || pathname === "/") ? url : pathname
}

export function configure(url: string, opts: Partial<ScaffoldOptions>, parent: NodeModule | null): ScaffoldOptions {

    const filename = opts.filename || getFilenameFromModule(parent)
    const casename = opts.casename || getCasenameFromUrl(url)

    if (filename === null) {
        throw new Error("Cannot determine Test Suite file name. Try specifiying it manually using the 'filename' option.")
    }

    return Object.assign({}, opts, { filename, casename, fallback: true })
}
