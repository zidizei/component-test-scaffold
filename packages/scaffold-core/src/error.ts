import path from "path"

export class LoadFromUrlError extends Error {

    name = "LoadFromUrlError"

    constructor(url: string) {
        super(`Failed to load Scaffold from URL '${url}'.`)
    }

}

export class LoadFromFileError extends Error {

    name = "LoadFromFileError"

    constructor(url: string, filePath: string) {
        super(`Failed to load Scaffold from URL '${url}' from file '${path.basename(filePath)}'.`)
    }

}
