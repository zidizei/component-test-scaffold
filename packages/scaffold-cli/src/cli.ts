import yargs from "yargs"

import { update } from "./index"
import version from "./version"

const CLI_VERSION = version()
const CLI_USAGE = "Usage: $0 [patterns..]"
const CLI_OPTS = {
    verbose: {
        demandOption: false,
        describe: "Show more details for updated scaffold files",
        type: 'boolean' as 'boolean',
    }
}

export type Parsed = {
    [x: string]: unknown
    verbose: boolean | undefined
    $0: string
}

export async function run(argv?: Array<string>) {
    argv = argv || process.argv.slice(2)

    const {_: patterns, ...parsed} = yargs(argv)
        .usage(CLI_USAGE)
        .options(CLI_OPTS)
        .version(CLI_VERSION)
        .alias("h", "help")
        .argv

    update(patterns, parsed)
}
