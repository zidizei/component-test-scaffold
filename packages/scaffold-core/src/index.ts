import https from "https"
import path from "path"

import { JSDOM } from "jsdom"

import { configure } from "./config"
import { loadUsingStrategy } from "./strategy"

export interface IScaffoldData {
    url: string
    test: string
    template: string
}

export interface IScaffoldResult {
    url: string
    template: string
    dom: JSDOM
}

export type Scaffold = {
    totalScaffolds: number
    currentScaffold: IScaffoldData
}

export type ScaffoldOptions = {
    filename: string
    casename: string
    writeback?: boolean
    loadStrategy?: "networkFirst" | "networkOnly" | "cacheFirst" | "cacheOnly" // TODO
    agent?: https.AgentOptions
}

export default async function (url: string, opts = {} as Partial<ScaffoldOptions>): Promise<IScaffoldResult> {
    const options = configure(url, opts, module.parent)

    const scaffoldPath = path.join(path.dirname(options.filename), "__scaffolds__")
    const scaffoldFile = `${path.basename(options.filename, path.extname(options.filename))}.scaffold.js`
    const scaffoldLocation = path.join(scaffoldPath, scaffoldFile)

    const html = await loadUsingStrategy(url, scaffoldLocation, options)
    const {template} = html.currentScaffold

    return {
        url,
        template,
        dom: new JSDOM(template, { url }),
    }
}
