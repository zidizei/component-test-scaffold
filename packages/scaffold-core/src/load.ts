import fs from "fs"
import https from "https"

import fetch from "node-fetch"
import prettier from "prettier"

import { Scaffold, ScaffoldOptions, IScaffoldData } from "."
import { LoadFromFileError, LoadFromUrlError } from "./error"

export async function loadFromUrl(url: string, filePath: string, options: ScaffoldOptions): Promise<Scaffold> {
    const {casename, filename} = options
    const agent = new https.Agent(options.agent)
    const resp = await fetch(url, { agent })

    if (!resp.ok) {
        throw new LoadFromUrlError(url)
    }

    const template = await resp.text()
    const scaffold: IScaffoldData = {
        url,
        filename,
        casename,
        template: prettier.format(template, { parser: "html" })
    }

    createOrUpdateFile(filePath, options, scaffold)

    return {
        totalScaffolds: 0,
        currentScaffold: scaffold,
    }
}

export async function loadFromFile(url: string, filePath: string, options: ScaffoldOptions): Promise<Scaffold> {
    const {casename} = options
    const scaffold: { [c: string]: IScaffoldData } = require(filePath)

    if (scaffold[casename] === undefined) {
        throw new LoadFromFileError(url, filePath)
    }

    return {
        totalScaffolds: Object.keys(scaffold).length,
        currentScaffold: scaffold[casename],
    }
}

export async function createOrUpdateFile(filePath: string, {casename, writeback}: ScaffoldOptions, data: IScaffoldData) {
    if (writeback !== true) { return }

    const scaffold = (fs.existsSync(filePath)) ? require(filePath) : {}
    scaffold[casename] = data

    let file = ``
    Object.keys(scaffold).forEach((scaffoldCase) => {
        file += `exports[\`${scaffoldCase}\`] = ${prettier.format(JSON.stringify(scaffold[scaffoldCase]), { parser: "json5" })}\n`
    })

    return new Promise((resolve, reject) => {
        fs.writeFile(filePath, file, (err) => {
            if (err) { reject(err) }
            resolve()
        })
    })
}
