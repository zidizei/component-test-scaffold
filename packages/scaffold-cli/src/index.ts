import load, { IScaffoldData } from "@component-test/scaffold-core"

import { Parsed } from "./cli"
import { glob } from "./util"

type Summary = {
    files: number
}

export async function update(patterns: Array<string>, {verbose}: Parsed): Promise<Summary> {
    const matches = await glob(patterns)
    const summary = {
        files: matches.length
    }

    await matches.forEach(async (filePath) => {
        const scaffold: { [c: string]: IScaffoldData } = require(filePath)
        const updates = Object.keys(scaffold).map((casename) => {
            const cas = scaffold[casename]

            if (verbose) {
                console.log("Updating '" + casename + "'.")
            }

            return load(cas.url, {
                filename: cas.test,
                casename,
                writeback: true,
                loadStrategy: "networkOnly"
            })
        })

        await Promise.all(updates)
    })

    return summary
}
