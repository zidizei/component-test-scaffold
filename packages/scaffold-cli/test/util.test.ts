import baseGlob from "glob"
import console from "console"

import { LoadFromUrlError } from "@component-test/scaffold-core/dist/error"

import { glob, log } from "../src/util"

describe("Scaffold CLI", () => {

    describe("Utils", () => {

        const mockedGlob = baseGlob as unknown as jest.Mock // Let TypeScript know that 'glob' has been mocked.

        beforeEach(() => {
            mockedGlob.mockRestore()
        })

        it("can glob scaffold files", async () => {
            mockedGlob.mockImplementation((_path, callback) => {
                callback(null, ["hello", "world"])
            })

            const actual = await glob([])
            expect(actual).toEqual(["hello", "world"])
            expect(mockedGlob).toHaveBeenCalledWith(
                expect.stringMatching(/__scaffolds__\/\*\.scaffold\.js$/),
                expect.any(Function)
            )
        })

        it("can glob scaffold files with patterns", async () => {
            mockedGlob.mockImplementation((_path, callback) => {
                callback(null, [])
            })

            await glob(["main.test", "other.spec"])
            expect(mockedGlob).toHaveBeenCalledWith(
                expect.stringMatching(/__scaffolds__\/\+\(main\.test\|other\.spec\)\.\*\.scaffold\.js$/),
                expect.any(Function)
            )
        })

        it("rejects with an error when failing to glob files", async () => {
            mockedGlob.mockImplementation((_path, callback) => {
                callback(new Error("Failed"), [])
            })

            await expect(glob([])).rejects.toThrow()
        })

    })

    describe("Logs", () => {

        let _stdout = ""
        let _stderror = ""
        let spyConsoleLog: jest.SpyInstance
        let spyConsoleError: jest.SpyInstance

        beforeEach(() => {
            _stdout = ""
            spyConsoleLog = jest.spyOn(console, "log")
            spyConsoleError = jest.spyOn(console, "error")
            spyConsoleLog.mockImplementation((msg) => {
                _stdout += msg
            })
            spyConsoleError.mockImplementation((msg) => {
                _stderror += msg
            })
        })

        afterEach(() => {
            jest.restoreAllMocks()
        })

        it("can log verbose messages", () => {
            log.verbose("test message")
            expect(spyConsoleLog).toHaveBeenCalledTimes(1)
            expect(_stdout).toMatchSnapshot()
        })

        it("can log info messages", () => {
            log.info("test message")
            expect(spyConsoleLog).toHaveBeenCalledTimes(1)
            expect(_stdout).toMatchSnapshot()
        })

        it("can log URL Errors", () => {
            log.error(new LoadFromUrlError("Message"))
            expect(spyConsoleError).toHaveBeenCalledTimes(2)
            expect(_stderror).toMatchSnapshot()
        })

    })

})
