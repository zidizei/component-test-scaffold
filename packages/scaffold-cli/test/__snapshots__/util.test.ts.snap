// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scaffold CLI Logs can log URL Errors 1`] = `
"[97m‚ö†Ô∏è Could not update Scaffold file(s).[39m
[97m[39m[31m[1mLoadFromUrlError[22m[39m
[31mFailed to load Scaffold from URL 'Message'.[39m"
`;

exports[`Scaffold CLI Logs can log info messages 1`] = `"[97mtest message[39m"`;

exports[`Scaffold CLI Logs can log verbose messages 1`] = `"[90mtest message[39m"`;
